### PR TITLE
refactor: add helper func to create cert and key

### PR DESCRIPTION
**Pull request by andrewrynhard**
It is common for us to create a certificate and key using a root CA.
This adds a helper function to make this process easier and
standardized. For now, the function is only used when generating
configs, but there are a number of places we should update in follow
ups.

### CODE REVIEW

File: pkg/config/types/v1alpha1/generate/generate.go [hunk 1]
```diff
@@ -33,10 +33,48 @@ const DefaultIPv6PodNet = "fc00:db8:10::/56"
 // DefaultIPv6ServiceNet is the network to be used for kubernetes Services when using IPv6-based master nodes
 const DefaultIPv6ServiceNet = "fc00:db8:20::/112"
 
-// CertStrings holds the string representation of a certificate and key.
-type CertStrings struct {
-	Crt string
-	Key string
+// Type represents a config type.
```
Comment (andrewrynhard) [inline pkg/config/types/v1alpha1/generate/generate.go]: This is just a move of the exported types to the top of the file.

File: pkg/config/types/v1alpha1/generate/generate.go [hunk 2]
```diff
@@ -131,98 +169,76 @@ type TrustdInfo struct {
 	Token string
 }
 
-func generateCertificateKey() (string, error) {
-	key, err := randBytes(32)
-	if err != nil {
-		return "", err
+func NewEtcdCA() (ca *x509.CertificateAuthority, err error) {
```
Comment (andrewrynhard) [inline pkg/config/types/v1alpha1/generate/generate.go]: I know there is a lot of commonality between the CA funcs, but they are explicit on purpose as a way to contain all related logic in a single function.

File: pkg/config/types/v1alpha1/generate/generate.go [hunk 3]
```diff
@@ -131,98 +169,76 @@ type TrustdInfo struct {
...
+		x509.RSA(false),
+		x509.Organization("talos"),
+		x509.NotAfter(time.Now().Add(87600 * time.Hour)),
 	}
 
-	return string(token), err
+	return x509.NewSelfSignedCertificateAuthority(opts...)
 }
 
-// genToken will generate a token of the format abc.123 (like kubeadm/trustd), where the length of the first string (before the dot)
```
Comment (andrewrynhard) [inline pkg/config/types/v1alpha1/generate/generate.go]: It is better to place unexpected functions at the bottom of the file.

File: pkg/config/types/v1alpha1/generate/generate.go [hunk 4]
```diff
@@ -388,46 +329,82 @@ func NewInput(clustername string, masterIPs []string, kubernetesVersion string)
...
+
+	tokenTemp[1], err = randBytes(lenSecond)
+	if err != nil {
+		return "", err
+	}
+
+	return tokenTemp[0] + "." + tokenTemp[1], nil
+}
+
+func isIPv6(addrs ...string) bool {
```
Comment (andrewrynhard) [inline pkg/config/types/v1alpha1/generate/generate.go]: This should be moved to our net package eventually

File: pkg/crypto/x509/x509.go [hunk 5]
```diff
@@ -538,6 +539,78 @@ func (p *PEMEncodedCertificateAndKey) MarshalYAML() (interface{}, error) {
 	return aux, nil
 }
 
+// NewCertficateAndKey generates a new key and certificate signed by a CA.
+func NewCertficateAndKey(crt *x509.Certificate, key interface{}, setters ...Option) (p *PEMEncodedCertificateAndKey, err error) {
```
Comment (andrewrynhard) [inline pkg/crypto/x509/x509.go]: This is the motivation for the PR.

File: pkg/config/types/v1alpha1/generate/generate.go (hunk 3 above)
Comment (smira) [reply pkg/config/types/v1alpha1/generate/generate.go]: I hope these are `unexported` :)

File: pkg/crypto/x509/x509.go [hunk 6]
```diff
@@ -538,6 +539,80 @@ func (p *PEMEncodedCertificateAndKey) MarshalYAML() (interface{}, error) {
...
+	} else {
+		priv, err = x509.ParseECPrivateKey(block.Bytes)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	csr, err := NewCertificateSigningRequest(priv, setters...)
+	if err != nil {
+		return nil, err
```
Comment (smira) [inline pkg/crypto/x509/x509.go]: errors in this function might benefit from any kind of wrapping (fmt.Errorf with %w or errors.Wrap), as sometimes it's hard to understand which part of the long flow returned an error

File: pkg/config/types/v1alpha1/generate/generate.go (hunk 3 above)
Comment (andrewrynhard) [reply pkg/config/types/v1alpha1/generate/generate.go]: lol yes

File: pkg/crypto/x509/x509.go (hunk 6 above)
Comment (andrewrynhard) [reply pkg/crypto/x509/x509.go]: Agree 👍

**Pull request merged**